packages/jest-runner/src/testWorker.ts

```diff
@@ -56,31 +58,41 @@ const getResolver = (
   // the test runner to the watch command. This is because jest-haste-map's
   // watch mode does not persist the haste map on disk after every file change.
   // To make this fast and consistent, we pass it from the TestRunner.
-  if (moduleMap) {
-    return Runtime.createResolver(config, moduleMap);
-  } else {
-    const name = config.name;
-    if (!resolvers[name]) {
-      resolvers[name] = Runtime.createResolver(
+  const name = config.name;
+  if (moduleMap || !resolvers.has(name)) {
+    resolvers.set(
+      name,
+      Runtime.createResolver(
         config,
-        Runtime.createHasteMap(config).readModuleMap(),
-      );
-    }
-    return resolvers[name];
+        moduleMap || Runtime.createHasteMap(config).readModuleMap(),
+      ),
+    );
   }
+  return resolvers.get(name)!;
 };
 
+const deserializedModuleMaps = new Map<string, number>();
 export async function worker({
   config,
   globalConfig,
   path,
   serializableModuleMap,
+  moduleMapUniqueID,
   context,
 }: WorkerData): Promise<TestResult> {
   try {
-    const moduleMap = serializableModuleMap
-      ? HasteMap.ModuleMap.fromJSON(serializableModuleMap)
-      : null;
+    // If the module map ID does not match what is currently being used by the
+    // config's resolver was passed, deserialize it and update the resolver.
+    let moduleMap: ModuleMap | null = null;
+    if (
+      serializableModuleMap &&
+      moduleMapUniqueID &&
```

Comment:
The unique ID starts from 0 so this check may be falsy for first creation of module map, isn't it?

Reply:
or start it at `1`

Reply:
Great catch!